src/miral/internal_client.cpp

```diff
@@ -99,14 +100,17 @@ class WlInternalClientRunner : public Base, public virtual InternalClientRunner
     std::thread thread;
     std::function<void(struct ::wl_display* display)> const client_code;
     std::function<void(std::weak_ptr<mir::scene::Session> const session)> const connect_notification;
+    std::optional<std::function<void()>> join_callback;
```

Comment:
It would be simpler (no need for an `optional<>` or branching logic) to always have a callback, but to default it to a NOP